GavinJoyce/ember-headlessui
# ✨ Implement `<Combobox />` component

###  ✨ What Changed and Why

This PR is a functionally complete implementation of a `<Combobox>` component based on https://headlessui.dev/react/combobox

Things that this PR implements:

* Show/hide options on mouse click
* Support for disabled component and individual options
* Support for a optional label element
* Full mouse support for focus and active elements
* Full keyboard support (`Enter`/`Space`/`ArrowUp`/`ArrowDown`/`ArrowLeft`/`ArrowRight`/`PageUp`/`PageDown`/`Home`/`End`/`Escape`)
* Support for searching the list using the keyboard
* Click outside the combobox to close
* Full accessibility implementation (`aria-expanded`/`aria-labelledby`/`aria-controls`/`aria-haspopup`/`aria-active-descendent`) and focus management
* Sample implementation in dummy app
* Multi-select

I ported the [React component tests for combobox](https://github.com/tailwindlabs/headlessui/blob/main/packages/%40headlessui-react/src/components/combobox/combobox.test.tsx) as Ember integration tests. 122 integration tests were ported. Some tests are marked as `skip` or `todo` because they are react specific or covered elesewhere.

### 😈 Before

Didn't exist.

### 😇 After

https://user-images.githubusercontent.com/70960245/163006210-6fccc1ed-af45-41d6-b04c-ebd453eeeeea.mov

### 🌈 Testing

1. Check out this branch locally
2. Visit http://localhost:4200/combobox/combobox-basic

Closes #138
